spring-data-mongodb/src/main/java/org/springframework/data/mongodb/core/MongoTemplate.java

```diff
@@ -847,6 +852,62 @@ public <T> List<T> find(Query query, Class<T> entityClass, String collectionName
 				new QueryCursorPreparer(query, entityClass));
 	}
 
+	@Override
+	public <T> CursorWindow<T> findWindow(CursorRequest cursorRequest, Query query, Class<T> entityType) {
+
+		Assert.notNull(entityType, "Entity type must not be null");
+
+		return findWindow(cursorRequest, query, entityType, getCollectionName(entityType));
+	}
+
+	@Override
+	// TODO: CursorRequest and Query declare both a Sort. Which one has precedence?
+	// CursorRequest is similar to Pageable in the sense of being required to define a sort order
+	public <T> CursorWindow<T> findWindow(CursorRequest cursorRequest, Query query, Class<T> entityType,
+			String collectionName) {
+		return doFindWindow(cursorRequest, query, entityType, entityType, collectionName);
+	}
+
+	<T> CursorWindow<T> doFindWindow(CursorRequest cursorRequest, Query query, Class<?> sourceClass, Class<T> targetClass,
+			String collectionName) {
+
+		Assert.notNull(cursorRequest, "CursorRequest must not be null");
+		Assert.notNull(query, "Query must not be null");
+		Assert.notNull(collectionName, "CollectionName must not be null");
+		Assert.notNull(sourceClass, "Entity type must not be null");
+		Assert.notNull(targetClass, "Target type must not be null");
+
+		ReadDocumentCallback<T> callback = new ReadDocumentCallback<>(mongoConverter, targetClass, collectionName);
+
+		if (cursorRequest instanceof OffsetCursorRequest offset) {
+
+			int limit = offset.getSize() + 1;
+
+			List<T> result = doFind(collectionName, createDelegate(query), query.getQueryObject(), query.getFieldsObject(),
+					sourceClass, new QueryCursorPreparer(query, new Query().with(cursorRequest.getSort()).getSortObject(), limit,
+							offset.getOffset(), sourceClass),
+					callback);
+
+			return CursorUtils.createWindow(cursorRequest, result);
+		}
+
+		if (cursorRequest instanceof KeysetCursorRequest keyset) {
+
+			int limit = keyset.getSize() + 1;
+
+			KeySetCursorQuery keysetPaginationQuery = CursorUtils.createKeysetPaginationQuery(query, keyset,
+					operations.getIdPropertyName(sourceClass));
+
+			List<T> result = doFind(collectionName, createDelegate(query), keysetPaginationQuery.query(),
```

Comment:
we're ignoring the `targetClass` here. Might be worth to call `doFind(createDelegate(query), collectionName, keysetPaginationQuery.query(), keysetPaginationQuery.fields(), sourceClass, targetClass, new QueryCursorPreparer(query, keysetPaginationQuery.sort(), limit, 0, sourceClass));` instead.

Reply:
`targetClass` is considered in `ReadDocumentCallback`. We might want to switch to `ProjectingReadCallback` though. Same for the imperative template.